src/peft/tuners/xlora/insertion.py

```diff
@@ -0,0 +1,284 @@
+        # scalings_layer = [batch_size, seq_len, n_classes]
+        scalings = scalings_layer[:, :, adapter].unsqueeze(-1)
+        # scalings_layer = [batch_size, seq_len, 1]
+        return x * scalings
+
+    def get_maybe_topk_scalings(self) -> torch.Tensor:
+        # xlora_scalings = [batch_size, seq_len, n_classes]
+        xlora_scalings: Tensor = self.model.internal_xlora_scalings[:, :, self.layer_number, :]  # type: ignore
+
+        if self.config.top_k_lora is not None:
+            _, topk_indices = torch.topk(xlora_scalings, k=self.config.top_k_lora, dim=1)
+
+            # Mask the topk to True, the rest to False
+            mask = torch.zeros_like(xlora_scalings, dtype=torch.bool)
+            mask.scatter_(1, topk_indices, True)
+
+            xlora_scalings = xlora_scalings * mask.to(xlora_scalings.dtype)
+
+        classifier: xLoRAClassifier = self.model.internal_xlora_classifier  # type: ignore
+        if classifier.config.enable_softmax_topk:
+            nonzero_mask = xlora_scalings != 0
+            softmax_res_nonzero = torch.softmax(xlora_scalings[nonzero_mask], dim=-1)
+            xlora_scalings[nonzero_mask] = softmax_res_nonzero
+
+        return xlora_scalings
+
+
+class xLoRALinearLayer(xLoRALayer):
+    def __init__(
+        self,
+        model: nn.Module,  # PeftModel
+        target: lora.Linear,
+        target_forward: Callable[..., Any],
+        layer_number: int,
+        config: xLoRAConfig,
+    ) -> None:
+        super().__init__(model, target, target_forward, layer_number, config)
+
+    def forward(self, x: Tensor, *args: Any, **kwargs: Any) -> Tensor:
+        """
+        This method is designed to be a drop-in-replacement for the peft LoRA layers' .forward method.
+        To use it, a bound method must be created (bound to an instance of the xLoRALayer class).
+        """
+
+        previous_dtype = x.dtype
+        xlora_scalings = self.get_maybe_topk_scalings()
+
+        if self.target.disable_adapters:
+            if self.target.merged:
+                self.target.unmerge()
+            result = self.target.base_layer(x, *args, **kwargs)
+        elif self.target.merged:
+            result = self.target.base_layer(x, *args, **kwargs)
+        else:
+            result = self.target.base_layer(x, *args, **kwargs)
+
+            for adapter_n, active_adapter in enumerate(self.target.active_adapters):
+                if active_adapter not in self.target.lora_A.keys():
+                    continue
+                lora_A = self.target.lora_A[active_adapter]
+                lora_B = self.target.lora_B[active_adapter]
+                dropout = self.target.lora_dropout[active_adapter]
+                scaling = self.target.scaling[active_adapter]
+                x = x.to(lora_A.weight.dtype)  # type: ignore
+                x_mod = self.apply_scalings_to_x(x, xlora_scalings, adapter_n)
```

Comment:
xLoRALinearLayer etc. are almost 1:1 copies of their respective LoRA layers. AFAICT only one line is added. I wonder if we could allow to use vanilla LoRA layers with a small refactor. This would avoid a lot of code duplication. Also, if we make improvements to the LoRA layers, it would be automatically reflected in XLoRA.

Reply:
Thank you for the great idea! I have pushed a commit which implements this.